### PR TITLE
Major refactor of ingest code to use OOP

### PR DESCRIPTION
**Pull request by brianhelba**

### CODE REVIEW

File: server/datasets/tcga/ingest.py [hunk 1]
```diff
@@ -25,392 +25,309 @@
...
-                                if limit and ingestCount >= limit:
-                                    global ingestUser
-                                    ingestUser = None
-                                    collectionCache.clear()
-                                    folderCache.clear()
-                                    return
-                                if progress is not None:
-                                    progress.update(
-                                        current=ingestCount,
-                                        total=limit if limit else None,
```
Comment (brianhelba) [inline server/datasets/tcga/ingest.py]: @manthey FYI, this was a bug. The `total` parameter must be an `int` or `float`.

**Comment by brianhelba** [comment]
@manthey When you have time, please take a brief look, and ensure you can run this locally. There's no need for a thorough code review at this point though.

**Comment by manthey** [comment]
It runs fine on my local environment.  A good improvement of the code.

A few minor things:

In Girder, you can set the minimum upload chunk size.  If it happens to be set > 32 MB, the ingest will fail.  To make the uploadWithProgress function, I copied the code from Girder directly, which should use the maximum of the `SettingKey.UPLOAD_MINIMUM_CHUNK_SIZE` and some reasonable value (32 MB currently, though for progress, it might be nice to have a small chunk so we get progress more often).  I can give you a simple fix for this, and I should submit a Girder fix for this, too.

In your extensively nested loops, you could now easily pull the core code to a separate function (so that the inner loop of `for dataTypePath in self._listAutoIndex(providerPath)[0]:` could call `self._ingestDataFromDataTypePath(dataTypePath)`, and that function would just contain that core (but now no longer nested an extra four layers).  This could be done at other loop levels instead or as well.

**Comment by zachmullen** [comment]
FWIW, I don't imagine that chunk size would ever be set > 32MB, I certainly would not recommend doing so. We chose 5MB as the default since that's the minimum chunk size allowed by S3.

That said, doing the check is certainly more robust than simply assuming :)

**Comment by brianhelba** [comment]
I'm going to merge now, just to keep the progress moving. 

@manthey Good point about the chunk size. 32MB was chosen for reading file-like objects off a local disk. We can go smaller when reading from an external HTTP download. I'll file an issue, and feel free to fix it here and in Girder, but it's low priority. 

As for the loops, I agree it could be even better (it would be so elegant in a functional language). Again,  we can fix later as a low priority.

**Pull request merged**